OCPBUGS-44632 Adding scoring strategy (#86820)
openshift/openshift-docs: modules/cnf-numa-resource-scheduling-strategies.adoc

// Module included in the following assemblies:
//

> // *scalability_and_performance/cnf-numa-aware-scheduling.adoc

Reviewer: Nit:
```suggestion
// * scalability_and_performance/cnf-numa-aware-scheduling.adoc
```

:_module-type: CONCEPT
[id="cnf-numa-resource-scheduling-strategies_{context}"]

> = NUMA Resource scheduling strategies 

Reviewer: Seems like it should be lowercase "resource" when it's a generic reference / not specifically referring to the "NUMA Resources Operator":
```suggestion
= NUMA resource scheduling strategies 
```

When scheduling high-performance workloads, the secondary scheduler can employ different strategies to determine which NUMA node within a chosen worker node will handle the workload. The supported strategies in {product-title} include `LeastAllocated`, `MostAllocated`, and `BalancedAllocation`. Understanding these strategies helps optimize workload placement for performance and resource utilization.

When a high-performance workload is scheduled in a NUMA-aware cluster, the following steps occur: 

. The scheduler first selects a suitable worker node based on cluster-wide criteria for example taints, labels, or resource availability.

. After a worker node is selected, the scheduler evaluates its NUMA nodes and applies a scoring strategy to decide which NUMA node will handle the workload.

> . Once a workload is scheduled, the selected NUMA node’s resources are updated to reflect the allocation.

Reviewer: Per IBM SG (for localization):
```suggestion
. After a workload is scheduled, the selected NUMA node’s resources are updated to reflect the allocation.
```

The default strategy applied is the `LeastAllocated` strategy. This assigns workloads to the NUMA node with the most available resources that is the least utilized NUMA node. The goal of this strategy is to spread workloads across NUMA nodes to reduce contention and avoid hotspots.

> .Example

Reviewer: Maybe for each of these three `.Example` titles describing the strategies in more detail, use a more descriptive title for each, e.g.:
```suggestion
.`LeastAllocated` example
```
or better yet, switch to an actual subsection heading (but maybe use `discrete` if you want to avoid showing up in the TOC):
```suggestion
[discrete]
[id="cnf-leastallocated-example_{context}"]
== LeastAllocated example
```
^ note lack of backticks if you do switch them all to `==` headings.

Assume a worker node has two NUMA nodes, and the workload requires 4 vCPUs and 8 GB of memory:

[cols="5,2,2,2,2,2", options="header"]
|===

> |NUMA Node |Total CPUs |Used CPUs |Total Memory (GB) |Used Memory (GB) |Available Resources

Reviewer: Sentence case for table headings as well:
```suggestion
|NUMA node |Total CPUs |Used CPUs |Total memory (GB) |Used memory (GB) |Available resources
```

> |NUMA 1 |16 |12 |64 |56 |4 CPUs, 8 GB Memory
> |NUMA 2 |16 |6 |64 |24 |10 CPUs, 40 GB Memory

Reviewer: ```suggestion
|NUMA 1 |16 |12 |64 |56 |4 CPUs, 8 GB memory
|NUMA 2 |16 |6 |64 |24 |10 CPUs, 40 GB memory
```

|===

> NUMA 2 has more available resources (10 CPUs, 40 GB memory) compared to NUMA 1 (4 CPUs, 8 GB memory) and therefore with the default strategy in place the workload is assigned to NUMA 2.

Reviewer: ```suggestion
NUMA 2 has more available resources (10 CPUs, 40 GB memory) compared to NUMA 1 (4 CPUs, 8 GB memory) and therefore, with the default strategy in place, the workload is assigned to NUMA 2.
```

> 

Reviewer: Blank line:
```suggestion
```

The `MostAllocated` strategy assigns workloads to the NUMA node with the least available resources that is the the most utilized NUMA node. The goal of this strategy is to consolidate workloads on fewer NUMA nodes, leaving other NUMA nodes idle or with free resources for energy efficiency or other workloads.

.Example
Using the same initial NUMA node states as earlier:

[cols="2,2"]
|===

> |NUMA Node |Available Resources
> |NUMA 1 |4 CPUs, 8 GB Memory
> |NUMA 2 |10 CPUs, 40 GB Memory

Reviewer: ```suggestion
|NUMA node |Available resources
|NUMA 1 |4 CPUs, 8 GB memory
|NUMA 2 |10 CPUs, 40 GB memory
```

|===

> NUMA 1 has fewer available resources compared to NUMA 2 and therefore with the `MostAllocated` strategy in place the workload is assigned to NUMA 1, further consolidating usage.

Reviewer: ```suggestion
NUMA 1 has fewer available resources compared to NUMA 2 and therefore, with the `MostAllocated` strategy in place, the workload is assigned to NUMA 1, further consolidating usage.
```

> The `BalancedAllocation` strategy assigns workloads to the NUMA node with the most balanced resource utilization across CPU and memory. The goal of this strategy is to avoid scenarios where one resource for example CPU is heavily utilized while another for example memory is underutilized.

Reviewer: ```suggestion
The `BalancedAllocation` strategy assigns workloads to the NUMA node with the most balanced resource utilization across CPU and memory. The goal of this strategy is to avoid scenarios where one resource, for example CPU, is heavily utilized while another for example memory is underutilized.
```

.Example
Assume a worker node has the following NUMA node states:

[cols="2,2,2,2"]
|===

> |NUMA Node |CPU Usage |Memory Usage |BalancedAllocation Score
> |NUMA 1 |60% |55% |High (More Balanced)
> |NUMA 2 |80% |20% |Low (Less Balanced)

Reviewer: ```suggestion
|NUMA node |CPU usage |Memory usage |`BalancedAllocation` score
|NUMA 1 |60% |55% |High (more balanced)
|NUMA 2 |80% |20% |Low (less balanced)
```

|===

> NUMA 1 has a more balanced CPU and memory utilization compared to NUMA 2 and therefore with the `BalancedAllocation` strategy in place the workload is assigned to NUMA 1. 

Reviewer: ```suggestion
NUMA 1 has a more balanced CPU and memory utilization compared to NUMA 2 and therefore, with the `BalancedAllocation` strategy in place, the workload is assigned to NUMA 1. 
```

The following table summarizes the different strategies and their outcomes:

[cols="2,3,3", options="header"]
|===
|Strategy |Description |Outcome

> |LeastAllocated |Favors NUMA nodes with the most available resources. |Spreads workloads to reduce contention and ensure headroom for high-priority tasks.
> |MostAllocated |Favors NUMA nodes with the least available resources. |Consolidates workloads on fewer NUMA nodes, freeing others for energy efficiency.
> |BalancedAllocation |Favors NUMA nodes with balanced CPU and memory usage. |Ensures even resource utilization, preventing skewed usage patterns.

Reviewer: Can the strategy names be marked up? Also, seems like the descriptions could all lose punctuation since they're all fragments.

```suggestion
|`LeastAllocated` |Favors NUMA nodes with the most available resources |Spreads workloads to reduce contention and ensure headroom for high-priority tasks
|`MostAllocated` |Favors NUMA nodes with the least available resources |Consolidates workloads on fewer NUMA nodes, freeing others for energy efficiency
|`BalancedAllocation` |Favors NUMA nodes with balanced CPU and memory usage |Ensures even resource utilization, preventing skewed usage patterns
```

. The scheduler first selects a suitable worker node based on cluster-wide criteria for example taints, labels, or resource availability.

. After a worker node is selected, the scheduler evaluates its NUMA nodes and applies a scoring strategy to decide which NUMA node will handle the workload.

. After a workload is scheduled, the selected NUMA node’s resources are updated to reflect the allocation.

The default strategy applied is the `LeastAllocated` strategy. This assigns workloads to the NUMA node with the most available resources that is the least utilized NUMA node. The goal of this strategy is to spread workloads across NUMA nodes to reduce contention and avoid hotspots.

[discrete]
[id="cnf-leastallocated-example_{context}"]
== LeastAllocated strategy (default) example
This strategy assigns workloads to the NUMA node with the most available resources, minimizing resource contention and spreading workloads across NUMA nodes. This reduces hotspots and ensures sufficient headroom for high-priority tasks. Assume a worker node has two NUMA nodes, and the workload requires 4 vCPUs and 8 GB of memory:

> 

Reviewer 2: ```suggestion
. Example initial NUMA nodes state
```
This table needs a title.

[cols="5,2,2,2,2,2", options="header"]
|===
|NUMA node |Total CPUs |Used CPUs |Total memory (GB) |Used memory (GB) |Available resources
|NUMA 1 |16 |12 |64 |56 |4 CPUs, 8 GB memory
|NUMA 2 |16 |6 |64 |24 |10 CPUs, 40 GB memory
|===

> NUMA 2 has more available resources (10 CPUs, 40 GB memory) compared to NUMA 1 (4 CPUs, 8 GB memory), therefore the workload is assigned to NUMA 2.

Reviewer 2: ```suggestion
Because NUMA 2 has more available resources compared with NUMA 1, the workload is assigned to NUMA 2.
```
I would keep this short and simple in case the table changes.

[discrete]
[id="cnf-mostallocated-example_{context}"]
== MostAllocated Strategy example

> This strategy consolidates workloads by assigning them to the NUMA node with the least available resources, which is the most utilized NUMA node. This approach helps free other NUMA nodes for energy efficiency or critical workloads requiring full isolation. Using the same initial NUMA node states as earlier:

Reviewer 2: ```suggestion
This strategy consolidates workloads by assigning them to the NUMA node with the least available resources, which is the most utilized NUMA node. This approach helps free other NUMA nodes for energy efficiency or critical workloads requiring full isolation. This example uses the "Example initial NUMA nodes state" values listed in the `LeastAllocated` section.
```

> 

Reviewer 2: ```suggestion
. Example NUMA nodes resources values
```

> [cols="2,2"]

Reviewer 2: ```suggestion
[cols="2,2", options="header"]
```
Personally, I don't see the need for this chunk of table here as it doesn't provide any meaningful information. But if it is going to be here, it is missing the header row bolding.

|===
|NUMA node |Available resources
|NUMA 1 |4 CPUs, 8 GB memory
|NUMA 2 |10 CPUs, 40 GB memory
|===

> The workload again requires 4 vCPUs and 8 GB memory. NUMA 1 has fewer available resources compared to NUMA 2, so the scheduler consolidates the workload on NUMA 1. The workload is assigned to NUMA 1, further utilizing its resources while leaving NUMA 2 idle or minimally loaded.

Reviewer 2: ```suggestion
The workload again requires 4 vCPUs and 8 GB memory. NUMA 1 has fewer available resources compared to NUMA 2, so the scheduler assigns the workload to NUMA 1, further utilizing its resources while leaving NUMA 2 idle or minimally loaded.
```
Is the existing workload being consolidated? Or does this mean the same thing as it is being assigned? If this is the same thing happening, this can be condensed for minimalism. Otherwise, I would specify that the existing workload is consolidated first, then the new one added.

[discrete]
[id="cnf-balanceallocated-example_{context}"]
== BalancedAllocation strategy example
This strategy assigns workloads to the NUMA node with the most balanced resource utilization across CPU and memory. The goal is to prevent imbalanced usage, such as high CPU utilization with underutilized memory. Assume a worker node has the following NUMA node states:

> 

Reviewer 2: ```suggestion
. Example NUMA nodes initial state for `BalancedAllocation`
```
Or something like that; needs a title.

> [cols="2,2,2,2"]

Reviewer 2: ```suggestion
[cols="2,2,2,2", options="header"]
```

|===
|NUMA node |CPU usage |Memory usage |`BalancedAllocation` score
|NUMA 1 |60% |55% |High (more balanced)
|NUMA 2 |80% |20% |Low (less balanced)
|===

NUMA 1 has a more balanced CPU and memory utilization compared to NUMA 2 and therefore, with the `BalancedAllocation` strategy in place, the workload is assigned to NUMA 1.  

> 

Reviewer 2: So table needs proper title; beyond that, I would give this summary **first** and make it a section with a heading.  Perhaps right above the LeastAllocated section. Much more scannable for advanced users who just need to reference the principles. You can provide the detailed breakdown after.

When scheduling high-performance workloads, the secondary scheduler can employ different strategies to determine which NUMA node within a chosen worker node will handle the workload. The supported strategies in {product-title} include `LeastAllocated`, `MostAllocated`, and `BalancedAllocation`. Understanding these strategies helps optimize workload placement for performance and resource utilization.

When a high-performance workload is scheduled in a NUMA-aware cluster, the following steps occur: 

.  The scheduler first selects a suitable worker node based on cluster-wide criteria. For example taints, labels, or resource availability.

. After a worker node is selected, the scheduler evaluates its NUMA nodes and applies a scoring strategy to decide which NUMA node will handle the workload.

. After a workload is scheduled, the selected NUMA node’s resources are updated to reflect the allocation.

The default strategy applied is the `LeastAllocated` strategy. This assigns workloads to the NUMA node with the most available resources that is the least utilized NUMA node. The goal of this strategy is to spread workloads across NUMA nodes to reduce contention and avoid hotspots.

> The following table summarizes the different strategies and their outcomes:

Reviewer 2: ```suggestion
```

[discrete]
[id="cnf-scoringstrategy-summary_{context}"]
== Scoring strategy summary

> 

Reviewer 2: ```suggestion
The following table summarizes the different strategies and their outcomes:
```
